### PR TITLE
[CDAP-16290] Added FLL instrumentation for Kinesis Streaming Source

### PR DESCRIPTION
**Pull request by vravish**

### CODE REVIEW

**Comment by vravish** [comment]
wait to merge https://github.com/cdapio/cdap/pull/12036 before this.

File: src/main/java/io/cdap/plugin/kinesis/streaming/KinesisStreamingSource.java [hunk 1]
```diff
@@ -59,6 +60,17 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
     config.validate(pipelineConfigurer.getStageConfigurer().getFailureCollector());
   }
 
+  @Override
+  public void prepareRun(StreamingSourceContext context) throws Exception {
+    super.prepareRun(context);
+
+    Schema schema = context.getInputSchema();
+    if (schema != null && schema.getFields() != null) {
+      recordLineage(context, config.referenceName, schema,
```
Comment (yaojiefeng) [inline src/main/java/io/cdap/plugin/kinesis/streaming/KinesisStreamingSource.java:69]: Let's put this method in ReferenceStreamingSource instead of cdap

Comment (vravish) [reply src/main/java/io/cdap/plugin/kinesis/streaming/KinesisStreamingSource.java:69]: so is it OK to duplicate the logic in different places?

Comment (yaojiefeng) [reply src/main/java/io/cdap/plugin/kinesis/streaming/KinesisStreamingSource.java:69]: This ReferenceStreamingSource should be in hydrator-common like other reference classes, that way it is not duplicating the logic

Comment (vravish) [reply src/main/java/io/cdap/plugin/kinesis/streaming/KinesisStreamingSource.java:69]: so should we move this class now or later? For now I just copied recordLineage to here

Comment (yaojiefeng) [reply src/main/java/io/cdap/plugin/kinesis/streaming/KinesisStreamingSource.java:69]: we can just keep it here for now.

File: src/main/java/io/cdap/plugin/kinesis/streaming/ReferenceStreamingSource.java [hunk 2]
```diff
@@ -58,4 +64,23 @@ public void run(DatasetContext datasetContext) {
     });
   }
 
+  /**
+   * Record field-level lineage for streaming source plugins (ReadOperation). This method should be called from
+   * prepareRun of any streaming source plugin.
+   * @param context StreamingSourceContext from prepareRun
```
Comment (yaojiefeng) [inline src/main/java/io/cdap/plugin/kinesis/streaming/ReferenceStreamingSource.java:75]: nit - put a new line before this

File: src/main/java/io/cdap/plugin/kinesis/streaming/ReferenceStreamingSource.java [hunk 3]
```diff
@@ -58,4 +64,23 @@ public void run(DatasetContext datasetContext) {
...
+   * @param outputName name of output dataset
+   * @param tableSchema schema of fields. Also used to determine list of field names. Schema and schema.getFields() must
+   * not be null.
+   * @param operationName name of the operation
+   * @param description operation description; complete sentences preferred
+   */
+  protected void recordLineage(StreamingSourceContext context, String outputName, Schema tableSchema,
+                               String operationName, String description) {
+    LineageRecorder lineageRecorder = new LineageRecorder(context, outputName);
+    lineageRecorder.createExternalDataset(tableSchema);
```
Comment (yaojiefeng) [inline src/main/java/io/cdap/plugin/kinesis/streaming/ReferenceStreamingSource.java]: streaming source does not need this, instead it should call https://github.com/cdapio/hydrator-plugins/blob/develop/spark-plugins/src/main/java/io/cdap/plugin/spark/FileStreamingSource.java#L74

File: src/main/java/io/cdap/plugin/kinesis/streaming/ReferenceStreamingSource.java [hunk 4]
```diff
@@ -58,4 +66,28 @@ public void run(DatasetContext datasetContext) {
...
+   * @param outputName name of output dataset
+   * @param tableSchema schema of fields. Also used to determine list of field names. Schema and schema.getFields() must
+   * not be null.
+   * @param operationName name of the operation
+   * @param description operation description; complete sentences preferred
+   */
+  protected void recordLineage(StreamingSourceContext context, String outputName, Schema tableSchema,
+                               String operationName, String description)
+    throws DatasetManagementException, TransactionFailureException {
+    Preconditions.checkNotNull(tableSchema, "schema for output %s is null.", outputName);
```
Comment (yaojiefeng) [inline src/main/java/io/cdap/plugin/kinesis/streaming/ReferenceStreamingSource.java]: Don't use Precondition, if will fail the pipeline. Should just do a if check and log as a warn message that field lineage will be missing

File: src/main/java/io/cdap/plugin/kinesis/streaming/ReferenceStreamingSource.java [hunk 5]
```diff
@@ -58,4 +68,34 @@ public void run(DatasetContext datasetContext) {
...
+   */
+  protected void recordLineage(StreamingSourceContext context, String outputName, Schema tableSchema,
+                               String operationName, String description)
+    throws DatasetManagementException, TransactionFailureException {
+    if (tableSchema == null) {
+      LOG.warn("Schema for output %s is null. Field-level lineage will not be recorded", outputName);
+      return;
+    }
+    if (tableSchema.getFields() == null) {
+      LOG.warn("schema.getFields() for output %s is null. Field-level lineage will not be recorded", outputName);
```
Comment (yaojiefeng) [inline src/main/java/io/cdap/plugin/kinesis/streaming/ReferenceStreamingSource.java]: nit - just say schema fields is empty

**Review by yaojiefeng — APPROVED** [review]
minor comment, LGTM

**Pull request merged**